### PR TITLE
Add ESDN mapping; nearly complete [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by anarchivist**
**DO NOT MERGE YET**

- [x] <strike>Currently this branch depends on the `dsl-resources`branch of Krikri (dpla/KriKri#61).</strike> **Merged**
- [x] <strike> One remaining gap is implementing the logic around how to parse multiple
  instances of `<dateCreated>` to implement the crosswalk listed in comments.</strike> **implemented**
- [ ] Needs discussion about how to harvest collection information. We can pull set specs in, but we probably need to improve the OAI parsers to use an optional mixin.
- [ ] Another gap has to do with error handling in the mapping DSL. The specs
  will currently fail given that the `record_uri` method won't work unless
  the OriginalRecord has been saved.

### CODE REVIEW

**Comment by anarchivist** [comment]
The latest commit introduces a dependency on dpla/KriKri#69.

**Comment by anarchivist** [comment]
Also, I need help (from @no-reply, probably) in figuring out how to deal with the `<dateCreated>` mapping.

**Comment by no-reply** [comment]
Can you clarify the behavior we're looking for?  I see a couple of places where there may be gaps, but I'm not sure which is actually the trouble, here.

**Comment by anarchivist** [comment]
@no-reply Basically, we may see something like ...

```
<originInfo>
  <dateCreated keyDate="yes">1949-05-30</dateCreated>
</originInfo>
```

... or ...

```
<originInfo>
  <dateCreated keyDate="yes" point="start">1945</dateCreated>
  <dateCreated point="end">1960</dateCreated>
</originInfo>
```

If I'm interpreting the crosswalk correctly, we want the date's `providedLabel` to be set to the value of  `<dateCreated keyDate="yes">` if there's only one (without the `@start` attribute), or the two values concatenated together if it's like the second example. 

We also need to parse the start and end dates if available to `edm:begin` and `edm:end`. 

I'm not sure how to move forward, but I see a few options:

* We could iterate using `:each => record.field('mods:originInfo')`, if we knew there were (reasonably) going to be one set of dates per record. (I don't know that; that might be a question for Gretchen.)
* We could yield multiple instances of `DPLA::MAP::TimeSpan`, which get addressed somehow in enrichment.
* This may represent a different set of needs for the DSL.

File: app/mappings/esdn_mods.rb [hunk 1]
```diff
@@ -1,4 +1,85 @@
...
+
+    contributor :class => DPLA::MAP::Agent, :each => record.field('mods:name').select { |name| name['mods:role'].map(&:value).include?('contributor') }, :as => :contrib do
+      providedLabel contrib.field('mods:namePart')
+    end
+
+    creator :class => DPLA::MAP::Agent, :each => record.field('mods:name').select { |name| name['mods:role'].map(&:value).include?('creator') }, :as => :creator_role do
+      providedLabel creator_role.field('mods:namePart')
+    end
+
+    date :class => DPLA::MAP::TimeSpan, :each => record.field('mods:originInfo', 'mods:dateCreated').select { |d| d.keyDate == "yes" }, :as => :created do
```
Comment (no-reply) [inline app/mappings/esdn_mods.rb]: I think you want: 
```ruby
date :class => DPLA::MAP::TimeSpan, :each => record.field('mods:originInfo'), :as => :created do
  providedLabel created.field('mods:dateCreated').match_attribute(:keyDate, 'yes')
  self.begin created.field('mods:dateCreated').match_attribute(:point, 'start')
  self.end created.field('mods:dateCreated').match_attribute(:point, 'end')
end
```

**Comment by no-reply** [comment]
> Another gap has to do with error handling in the mapping DSL. The specs will currently fail given that the record_uri method won't work unless the OriginalRecord has been saved.

Is this a gap, or just a problem with the spec.  You might want to just create a saved (`#exists? # => true`) `OriginalRecord`.

**Comment by anarchivist** [comment]
Would that be something like a `before` condition in the spec here?

**Comment by no-reply** [comment]
That would work. It is probably also something that we will want to do often enough to merit a factory approach.  Switching `build` for `create` might already work.  If not, a new build strategy may be called for: https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#using-factories

**Comment by no-reply** [comment]
> Needs discussion about how to harvest collection information. We can pull set specs in, but we probably need to improve the OAI parsers to use an optional mixin.

Proposal: the OAI parsers get a `header` node, via mixin. `setSpec` will be available under this.  Translations from OAI setSpecs to collection names is declared a mapping issue and we consider options for building support into the DSL.

**Comment by anarchivist** [comment]
> Proposal: the OAI parsers get a `header` node, via mixin. `setSpec` will be available under this.  Translations from OAI setSpecs to collection names is declared a mapping issue and we consider options for building support into the DSL.

Added as [#87448174](https://www.pivotaltracker.com/story/show/87448174).

**Comment by anarchivist** [comment]
Superseded by dpla/heidrun-mappings#2.

**Pull request closed without merging**